pkg/template/processor.go

```diff
@@ -0,0 +1,193 @@
+	for param, val := range values {
+		v := templateprocessing.GetParameterByName(tmpl, param)
+		if v != nil {
+			v.Value = val
+			v.Generate = ""
+		}
+	}
+	// convert the template into a set of objects
+	tmplProcessor := templateprocessing.NewProcessor(map[string]generator.Generator{
+		"expression": generator.NewExpressionValueGenerator(rand.New(rand.NewSource(time.Now().UnixNano()))),
+	})
+	if err := tmplProcessor.Process(tmpl); len(err) > 0 {
+		return nil, errs.Wrap(err.ToAggregate(), "unable to process template")
+	}
+	var result templatev1.Template
+	if err := p.scheme.Convert(tmpl, &result, nil); err != nil {
+		return nil, errs.Wrap(err, "failed to convert template to external template object")
+	}
+	return Filter(result.Objects, filters...), nil
+}
+
+// Apply applies the objects, ie, creates or updates them on the cluster
+func (p Processor) Apply(objs []runtime.RawExtension) error {
+	for _, rawObj := range objs {
+		obj := rawObj.Object
+		if obj == nil {
+			continue
+		}
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		_, err := p.ApplySingle(obj, true, nil)
+		if err != nil {
+			return errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
+		}
+	}
+	return nil
+}
+
+// ApplySingle creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then based on the UpdateStrategy it's updated or let as it is and updated is skipped.
+// The return boolean says if the object was either created or updated
+func (p Processor) ApplySingle(obj runtime.Object, forceUpdate bool, owner v1.Object) (bool, error) {
+	gvk := obj.GetObjectKind().GroupVersionKind()
+	createdOrUpdated, err := p.createOrUpdateObj(obj, forceUpdate, owner)
+	if err != nil {
+		return createdOrUpdated, errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
+	}
+	return createdOrUpdated, nil
+}
+
+func (p Processor) createOrUpdateObj(newResource runtime.Object, forceUpdate bool, owner v1.Object) (bool, error) {
+	// gets the meta accessor to the new resource
+	metaNew, err := meta.Accessor(newResource)
+	if err != nil {
+		return false, errors.Wrapf(err, "cannot get metadata from %+v", newResource)
+	}
+
+	// creates a deepcopy of the new resource to be used to check if it already exists
+	existing := newResource.DeepCopyObject()
+
+	// set current object as annotation
+	//newWithoutAnnotation := newResource.DeepCopyObject()
+	annotations := metaNew.GetAnnotations()
+	newConfiguration := getNewConfiguration(newResource)
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	annotations[LastAppliedConfigurationAnnotationKey] = newConfiguration
+	metaNew.SetAnnotations(annotations)
+
+	// gets current object (if exists)
+	namespacedName := types.NamespacedName{Namespace: metaNew.GetNamespace(), Name: metaNew.GetName()}
+	if err := p.cl.Get(context.TODO(), namespacedName, existing); err != nil {
```

Comment:
shouldn't we check that the error was `IsNotFound` after the call to `Get` and before the call to `Create`?

Reply:
well, I had it originally in the first version of my implementation, but then I realized why it should try to create the resource anyway - no matter what was the reason for the error. 
I do remember that in the tenant service I got an error of missing RBAC for resources that were not found. However, I'm not sure if that was because of OpenShift 3, because I used http requests instead of the go client or because of any other reason...

Reply:
I'm not sure about that particular case you got in tenant service and in Kube in general but it's considered to be a good practice to return Auth errors instead of 404 (or any other NotFound error) in case of missing resources for clients which doesn't have permissions to read the resource. It's done for security reasons so if you don't have permissions to read resources you won't be able to check if the resource exists or not.

Reply:
But if the client does have enough permissions but still gets RBAC errors just because the resource is missing then it looks like a bug :)

Reply:
Yes, that was exactly the case when I got the auth error n tenant service, but the user we use in there should have enough permissions. However, it could be a different behavior for users and SA. 
Anyway, I changed the implementation back to the original version so it's trying to create the resource only when it gets "NotFound" - see 914a786

Reply:
ok, thanks @MatousJobanek